### ISSUE TITLE
Add option for no-unused-expressions

### ISSUE THREAD

**Issue by ethanresnick** [eslint/eslint#2102]
I'm using [Chai's expect](http://chaijs.com/api/bdd/) syntax for testing, and I'm getting `no-unused-expression` errors. I understand that this is because expressions like `expect(...).to.be.undefined` look like property access, even though chai's actually using the property's getter to call a method behind the scenes that causes some side effects.

I did a bit of googling and at least a [few](https://github.com/roadhump/SublimeLinter-eslint/issues/9) [other](https://github.com/gziolo/es6-browser-boilerplate/issues/1) [people](http://stackoverflow.com/questions/27533471/unit-tests-with-chai-and-eshint-are-not-passing) have encountered this.

I'm not sure what the best solution is (I don't know enough about parsing JS to say) but it seems like one possibility might be an option that would not flag property access chains that start with certain identifiers. So the user could specify `"expect"` as the option's value and then all these chai calls would be let through.

**Comment by michaelficarra** [comment]
Just turn the rule off, at least for your test directory. This is a pretty explicit violation of exactly what this rule is trying to prevent.

**Comment by nzakas** [comment]
@ethanresnick can you please provide a sample code snippet and the exact console output that ESLint is giving you?

On the surface, I'm inclined to agree with @michaelficarra that your best option is to disable the rule in your tests, where it's arguably less useful.

**Comment by ethanresnick** [comment]
Sure!

Here's the code, in the most reduced form:

```
var expect = require('chai').expect;

describe("the number 4", function() {
  it("should be truthy", function() {
    expect(4).to.be.true;
  });
});
```

And the console output:

```
5:4  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions

✖ 1 problem (1 error, 0 warnings)
```

With the config:
```
{
  "rules": {
    "strict": 0,
    "curly": 0,
    "indent": [2, 2, {indentSwitchCase: true}],
    "brace-style": [2, "stroustrup", { "allowSingleLine": true }],
    "quotes": 0
  },
  "env": {
    "node": true,
    "mocha": true
  }
}
```

As far as just turning the expression off goes: yes, that's always an option. But, of course, if there's a way to leave it on to get whatever other utility it provides, while telling it not to flag these usages that are beyond my control, that would be nice :)

**Comment by devangnegandhi** [comment]
I am facing this exact same issue. I currently use the `eslint-disable` option as a workaround. But an option to configure this would be awesome!

**Comment by qmmr** [comment]
:+1: 
It looks like chai's `expect` is the most common exception from this rule as it usually eats expressions.

**Comment by nzakas** [comment]
Yup, we are aware. We still don't have a solution that doesn't involve building-in knowledge of Chai. If we can't find one soon, we are going to just recommend you turn off the rule for tests.

**Comment by a-c-m** [comment]
Hitting this same issue. Would love a work around, perhaps to exclude that rule from a pattern e.g. *.spec.js (apologies if this is possible already, i could not work out how).

**Comment by michaelficarra** [comment]
@a-c-m it is possible through multiple eslint invocations

**Comment by nzakas** [comment]
Sorry all, we still don't have a way of resolving this without baking-in knowledge of Chai. As a general principle, we do not put library-specific code in ESLint.

While it's unfortunate and not ideal, it's up to you to turn this off where appropriate. You can do that in a config or using online comments. See our docs for more: http://eslint.org/docs/user-guide/configuring

I'm closing this issue as we don't plan on addressing this.

**Comment by LFDM** [comment]
FYI, a small chai plugin which gives you these property getters as functions: https://www.npmjs.com/package/chai-lint

**Comment by justin-lau** [comment]
[dirty-chai](https://github.com/prodatakey/dirty-chai) works for me. It also automatically converts plugin assertions.

**Comment by niftylettuce** [comment]
Thanks @justin-lau

**Comment by StephenGrider** [comment]
@justin-lau dirty-chai addressed this issue fantastically for me, thanks for sharing

**Comment by niftylettuce** [comment]
:+1: for `dirty-chai`

```bash
npm install --save-dev dirty-chai
```

```diff
var chai = require('chai');
+// Load dirty chai first to hook plugin extensions
+var dirtyChai = require('dirty-chai');
+chai.use(dirtyChai);
```

**Comment by screendriver** [comment]
👍 for [dirty-chai](https://github.com/prodatakey/dirty-chai)

**Comment by djfm** [comment]
Might be a lot of work for not a lot of gain, but wouldn't a rule that ignores "no-unused-expressions" when "env" is set to "mocha" and only for the few mocha magic properties (ok, undefined, ...) make sense?

**Comment by platinumazure** [comment]
> Might be a lot of work for not a lot of gain, but wouldn't a rule that ignores "no-unused-expressions" when "env" is set to "mocha" and only for the few mocha magic properties (ok, undefined, ...) make sense?

In core ESLint, no, but you are absolutely welcome to write a plugin rule which emulates no-unused-expressions except in the circumstances you described.

**Comment by djfm** [comment]
Ok maybe some day when I'm really bored :) After a little more thought I don't think it's worth it or even conceptually such a good idea, eslint-disable-next-line is not so hard to write and provides more information to the reader.

**Comment by AMorgaut** [comment]
@nzakas with all respect, you said:

> Sorry all, we still don't have a way of resolving this without baking-in knowledge of Chai. As a general principle, **we do not put library-specific code in ESLint**.

But actually, ESLint, as JSHint, does support "environment options" such as `jasmine`, `mocha`, and `qunit`, plus `protractor` and `atomtest` on ESLint side... so I think that adding a `chai` environment option would not break existing ESLint library approach.

I just wrote a [related feature request to the JSHint project](https://github.com/jshint/jshint/issues/3070)

As I described in that feature request, another non 'library related' approach could be to add a `expr: 'accessor'` sub-option to not raise the warning when we ends an expression with the access to an `accessor property` (I mentioned use cases like forced singleton instantiation). It can be interesting but:

* would probably be more complicated to support than just supporting the chai use case
* could hide expected warnings if we only want it on the chai API (such sub option is probably more legitimate to be activated at line level tahn at file level)

**Comment by mattdell** [comment]
This eslint plugin seems to work for me: https://www.npmjs.com/package/eslint-plugin-chai-friendly

I agree an environment option, similar to mocha, is the best way forward.

**Comment by not-an-aardvark** [comment]
@AMorgaut

> But actually, ESLint, as JSHint, does support "environment options" such as `jasmine`, `mocha`, and `qunit`, plus `protractor`, `atomtest`, and `embertest` on ESLint side... so I think that adding a `chai` environment option would not break existing ESLint library approach.

[Environments](http://eslint.org/docs/user-guide/configuring#specifying-environments) in ESLint are used to define global variables, and occasionally set parser options. By design, they cannot change the behavior of rules like `no-unused-expressions`, so it wouldn't be possible to use an environment in the way that you're suggesting.

Additionally, while ESLint used to add environments for custom libraries, it [no longer does this](https://github.com/sindresorhus/globals/issues/82). 

> As I described in that feature request, another non 'library related' approach could be to add a `expr: 'accessor'` sub-option to not raise the warning when we ends an expression with the access to an `accessor property`(cf [ES5 Property Attributes](http://www.ecma-international.org/ecma-262/5.1/#sec-8.6.1)) . I mentioned related use cases like forced singleton instantiation. 

ESLint is a static analysis tool, so it's not possible to tell whether any given property access triggers a getter. Providing a list of accessor properties via configuration might be feasible; https://github.com/eslint/eslint/issues/8075 is another option.

**Comment by AMorgaut** [comment]
@not-an-aardvark Thank for your feedback

Even if it's a bit frustrating, I've seen in the meantime, and understand, that supporting other environment configs are not exactly the chosen path for either JSHint or ESLint. Probably is it better to get a list of environment lint plugins so they could be co-supported by some of the target platform contributors.

> ESLint is a static analysis tool, so it's not possible to tell whether any given property access triggers a getter. 

Well, I started to work on AST transforms. As far as I saw both [babylon](https://github.com/babel/babylon) & [Espree](https://github.com/eslint/espree) (used by ESLint) being based on [Acorn](https://github.com/ternjs/acorn), it looks quite possible from a static analysis tool to see when a property is defined via 

* `Object.defineProperty(context, "name", {get: func});`
* or `context = {get foo: func};`
* or `class Context { get foo() {} }`

But yes there is a limitation when working in cross file contexts...

> Providing a list of accessor properties via configuration is feasible

It is the direction that might be taken on the JSHint side [as mentioned on this proposal](https://github.com/jshint/jshint/issues/3086)

**Comment by artforlife** [comment]
I am having a similar issue in the form of getting the following error: 

> 'should' is assigned a value but never used.

It comes from the following line:

`var should = chai.should()`

I have tried `dirty-chai` to no avail. How could I deal with this error?

**Comment by aaron-goshine** [comment]
@nzakas you are one of the expert amongst us, What are the advantages in chai implementation, to omit the useage of the  parentheses `()`
-- it seems like it might be causing more problem that it solve..

**Comment by michaelficarra** [comment]
@aaron-goshine You're absolutely right. The advantage is no parentheses. The disadvantage is mass confusion and many broken assumptions. Nobody should ever have side-effecting getters, without exception.
